sdk/iot/Azure.Iot.Hub.Service/src/Authentication/SasTokenAuthenticationPolicy.cs

```diff
@@ -0,0 +1,66 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core;
+using Azure.Core.Pipeline;
+
+namespace Azure.Iot.Hub.Service.Authentication
+{
+    /// <summary>
+    /// The shared access signature based HTTP pipeline policy.
+    /// This authentication policy injects the sas token into the HTTP authentication header for each HTTP request made.
+    /// </summary>
+    internal class SasTokenAuthenticationPolicy : HttpPipelinePolicy
+    {
+        private readonly TimeSpan _getTokenTimeout = TimeSpan.FromSeconds(30);
+        private readonly TokenCredential _credential;
+
+        internal SasTokenAuthenticationPolicy(TokenCredential credential)
+        {
+            _credential = credential;
+        }
+
+        public override void Process(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline)
+        {
+            ProcessAsync(message, pipeline, false).EnsureCompleted();
+        }
+
+        public override ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline)
+        {
+            return ProcessAsync(message, pipeline, true);
+        }
+
+        private async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
+        {
+            if (message.Request.Uri.Scheme != Uri.UriSchemeHttps)
+            {
+                throw new InvalidOperationException("Token authentication is not permitted for non TLS protected (https) endpoints.");
+            }
+
+            await AddHeaders(message, async).ConfigureAwait(false);
+
+            if (async)
+            {
+                await ProcessNextAsync(message, pipeline).ConfigureAwait(false);
+            }
+            else
+            {
+                ProcessNext(message, pipeline);
+            }
+        }
+
+        private async Task AddHeaders(HttpMessage message, bool async)
```

Comment:
isAsync? useAsync?

Reply:
The track 2 libraries actually recommend that it be called "async", not sure why.

Reply:
https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs#L54

Reply:
@KrzysztofCwalina @weshaggard 
This seems weird to me.

1. I thought the general guidance for boolean naming was to prefix it with a verb to indicate an either/or.
2. In this case, `async` is also a keyword in c# which makes the subsequent code a bit confusing.